### PR TITLE
SONARJAVA-2078 Edges of Exploded Graph should be modeled explicitly

### PR DESCRIPTION
**Pull request by saberduck**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/se/ProgramState.java [hunk 1]
```diff
@@ -417,4 +419,37 @@ public SymbolicValue exitValue() {
...
+      BinarySymbolicValue binarySymbolicValue = (BinarySymbolicValue) sv;
+      addLearnedConstraint(result, binarySymbolicValue.getLeftOp(), null);
+      addLearnedConstraint(result, binarySymbolicValue.getRightOp(), null);
+    }
+  }
+
+  Set<LearnedAssociation> learnedAssociations(ProgramState parent) {
+    ImmutableSet.Builder<LearnedAssociation> result = ImmutableSet.builder();
+    values.forEach((s, sv) -> {
+      if (parent.getValue(s) != sv) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/ProgramState.java]: MAJOR Change this comparison to use the equals method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

File: java-frontend/src/main/java/org/sonar/java/se/ProgramState.java [hunk 2]
```diff
@@ -417,4 +419,37 @@ public SymbolicValue exitValue() {
...
+      if (parentConstraint == null || !parentConstraint.equals(childConstraint)) {
+        addLearnedConstraint(result, sv, childConstraint);
+      }
+    });
+    return result.build();
+  }
+
+  private static void addLearnedConstraint(ImmutableSet.Builder<LearnedConstraint> result, SymbolicValue sv, Constraint c) {
+    result.add(new LearnedConstraint(sv, c));
+    // FIXME this might end up adding twice the same SV in learned constraints. Safe because null constraints are filtered anyway
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/ProgramState.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134)

File: java-frontend/src/main/java/org/sonar/java/se/ProgramState.java [hunk 3]
```diff
@@ -417,4 +419,37 @@ public SymbolicValue exitValue() {
...
+    // FIXME remove null usage
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/ProgramState.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: java-frontend/src/main/java/org/sonar/java/se/LearnedConstraint.java [hunk 4]
```diff
@@ -46,6 +49,22 @@ public Constraint constraint() {
   }
 
   @Override
+  public boolean equals(Object o) {
+    if (this == o)
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/LearnedConstraint.java]: MINOR Missing curly brace. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00121)

File: java-frontend/src/main/java/org/sonar/java/se/LearnedConstraint.java [hunk 5]
```diff
@@ -46,6 +49,22 @@ public Constraint constraint() {
...
+      return true;
+    if (o == null || getClass() != o.getClass())
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/LearnedConstraint.java]: MINOR Missing curly brace. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00121)

File: java-frontend/src/main/java/org/sonar/java/se/LearnedAssociation.java [hunk 6]
```diff
@@ -40,6 +42,22 @@ public SymbolicValue symbolicValue() {
   }
 
   @Override
+  public boolean equals(Object o) {
+    if (this == o)
+      return true;
+    if (o == null || getClass() != o.getClass())
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/LearnedAssociation.java]: MINOR Missing curly brace. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00121)

File: java-frontend/src/main/java/org/sonar/java/se/LearnedAssociation.java [hunk 7]
```diff
@@ -40,6 +42,22 @@ public SymbolicValue symbolicValue() {
   }
 
   @Override
+  public boolean equals(Object o) {
+    if (this == o)
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/LearnedAssociation.java]: MINOR Missing curly brace. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00121)

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 4 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: java-frontend/src/main/java/org/sonar/java/se/LearnedAssociation.java (hunk 6 above)
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/LearnedAssociation.java]: :+1:

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: java-frontend/src/main/java/org/sonar/java/se/ExplodedGraph.java [hunk 8]
```diff
@@ -156,7 +126,55 @@ public String toString() {
...
+
+    Set<LearnedAssociation> learnedAssociations() {
+      if (la == null) {
+        la = child.programState.learnedAssociations(parent.programState);
+      }
+      return la;
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/ExplodedGraph.java]: unsure you actually need to implement hashcode/equals for edges. IMO those should be dropped for the time being.

File: java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java [hunk 9]
```diff
@@ -117,17 +117,17 @@ private void run(@Nullable final ExplodedGraph.Node node, @Nullable final Symbol
...
         if (learnedConstraints.anyMatch(terminateTraversal)) {
           continue;
         }
+      } else {
+        newTrackSymbol = nodeSymbol.trackSymbol;
       }
-      for (ExplodedGraph.Node parent : currentNode.getParents()) {
-        workList.push(new NodeSymbol(parent, newTrackSymbol));
-      }
+      currentNode.parents().forEach(parent -> workList.push(new NodeSymbol(parent, newTrackSymbol)));
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java]: As far as I can see this is a cosmetic refactoring and should not be  should not be part of this commit.

File: java-frontend/src/main/java/org/sonar/java/se/LearnedConstraint.java [hunk 10]
```diff
@@ -25,13 +25,16 @@
...
+import java.util.Objects;
+
 public class LearnedConstraint {
   final SymbolicValue sv;
 
   @Nullable
   final Constraint constraint;
 
   public LearnedConstraint(SymbolicValue sv, @Nullable Constraint constraint) {
+    Objects.requireNonNull(sv);
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/LearnedConstraint.java]: unrelated to this commit.

File: java-frontend/src/test/java/org/sonar/java/se/CheckerDispatcherTest.java [hunk 11]
```diff
@@ -49,7 +49,7 @@ private static ExplodedGraphWalker mockExplodedGraphWalker() {
     ExplodedGraphWalker explodedGraphWalker = mock(ExplodedGraphWalker.class);
     explodedGraphWalker.programPosition = new ProgramPoint(new CFG.Block(1));
     explodedGraphWalker.programState = mock(ProgramState.class);
-    explodedGraphWalker.node = new ExplodedGraph.Node(explodedGraphWalker.programPosition, explodedGraphWalker.programState);
+    explodedGraphWalker.node = new ExplodedGraph().node(explodedGraphWalker.programPosition, explodedGraphWalker.programState);
```
Comment (benzonico) [inline java-frontend/src/test/java/org/sonar/java/se/CheckerDispatcherTest.java]: should be part of your first commit. irrelevant to the third one.

File: java-frontend/src/test/java/org/sonar/java/se/ExplodedGraphTest.java [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+import org.junit.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class ExplodedGraphTest {
+
+  @Test
+  public void test_child_parent() {
```
Comment (benzonico) [inline java-frontend/src/test/java/org/sonar/java/se/ExplodedGraphTest.java]: not really sure from the test name what is actually tested.

File: java-frontend/src/main/java/org/sonar/java/se/ExplodedGraph.java (hunk 8 above)
Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/java/se/ExplodedGraph.java]: dropped for now

File: java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java (hunk 9 above)
Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java]: reverted refactoring

File: java-frontend/src/test/java/org/sonar/java/se/CheckerDispatcherTest.java (hunk 11 above)
Comment (saberduck) [inline java-frontend/src/test/java/org/sonar/java/se/CheckerDispatcherTest.java]: I disagree, this change is related to the fact that Node constructor is now private.

Comment (benzonico) [inline java-frontend/src/test/java/org/sonar/java/se/CheckerDispatcherTest.java]: which should probably be part of your first commit.

Comment (saberduck) [inline java-frontend/src/test/java/org/sonar/java/se/CheckerDispatcherTest.java]: OK, I rearranged commits to include this change in the first one

**Pull request merged**